### PR TITLE
‼️ BREAKING: Compare `Dict` nodes by content

### LINKED ISSUE #1917: Distinct instances of the same node do not compare equal
Example input:
```python
n1 = load_node(42)
n2 = load_node(42)
print n1, type(n1)
print n2, type(n2)
print n1 == n2
```
Output:
```
uuid: ddf9a6de-56bf-4783-96d2-33400df20f43 (pk: 42) <class 'aiida.orm.data.upf.UpfData'>
uuid: ddf9a6de-56bf-4783-96d2-33400df20f43 (pk: 42) <class 'aiida.orm.data.upf.UpfData'>
False
```
Tested with types `UpfData`, `WorkCalculation`, `PwCalculation`.

EDIT: also tested with more "basic" types `aiida.orm.data.str.Str` and `aiida.orm.data.int.Int`. For these, some more discussion is possibly required: should the PK/UUID be compared, or the actual data themselves?

### PR DESCRIPTION
**Pull request by mbercx**
Fixes #1917

Currently there is an inconsistency in how the base data type node
instances compare equality. All base types compare based on the content
of the node, whereas `Dict` instances rely on the UUID fallback
introduced in #4753. After a long discussion started by #1917, it was
finally decided that the best way forward is to make the equality
comparison consitent among the base types (see #5187).

Here we adapt the `__eq__` method of the `Dict` class to compare
equality by content instead of relying on the fallback comparison of
the UUIDs.

### CODE REVIEW

**Review by muhrin — DISMISSED** [review]
The code changes look fine to me and given that the conclusion of the related discussions was that this breaking change could go ahead I think it's fine to merge.

**Comment by mbercx** [comment]
Thanks @muhrin! I'll leave the merging to coding week merge master @chrisjsewell.

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @mbercx . See comment of removing `__ne__`. Would do same for other base types that have just plain negation of `__eq__`. Then I think this OK to merge. Final important thing would be to [update the wiki page for v2.0 migration](https://github.com/aiidateam/aiida-core/wiki/AiiDA-2.0-plugin-migration-guide) and make a note of this important change.

File: aiida/orm/nodes/data/dict.py [hunk 1]
```diff
@@ -71,10 +71,12 @@ def __setitem__(self, key, value):
...
     def __eq__(self, other):
-        if isinstance(other, dict):
-            return self.get_dict() == other
+        if isinstance(other, Dict):
+            return self.get_dict() == other.get_dict()
+        return self.get_dict() == other
 
-        return super().__eq__(other)
+    def __ne__(self, other):
+        return not self == other
```
[the remark this answers is not in the document]
Comment (mbercx) [reply aiida/orm/nodes/data/dict.py:79]: Thanks @sphuber! Wasn't aware this was the default behaviour, will remove it for `List` and `BaseType` as well then.

**Comment by mbercx** [comment]
Thanks for the review @sphuber! I've removed the `_ne_` operator for the base types and have added a description of this backwards-incompatible change to the v2.0 plugin migration guide:

https://github.com/aiidateam/aiida-core/wiki/AiiDA-2.0-plugin-migration-guide#equality-comparison-of-dict-nodes

**Comment by sphuber** [comment]
So can this close #1917? @mbercx ? Any other issues that this PR can close?

**Comment by mbercx** [comment]
> So can this close #1917? @mbercx ? 

Yes, I would say so. It might not be the last time we discuss how to compare node equality, but I think we've all settled on "base types compare content all others compare UUID by default" for now. I'll update the OP.

> Any other issues that this PR can close?

Not afaik.

**Review by sphuber — APPROVED** [review]

**Pull request merged**